labelbox_dev/task.py

```diff
@@ -0,0 +1,186 @@
+
+    @abstractmethod
+    def wait_until_done(self):
+        ...
+
+
+class Task(BaseTask, ABC):
+    """ Abstract class representing an asynchronous task with single input.
+    Users can call `task.result` and `task.error` to obtain either the
+    job result upon success, or job error upon failure, respectively.
+    """
+
+    def __init__(self, json):
+        super().__init__(json)
+
+    @property
+    @abstractmethod
+    def result(self):
+        ...
+
+    @property
+    @abstractmethod
+    def error(self):
+        ...
+
+
+class BulkTask(BaseTask, ABC):
+    """ Abstract class representing an asynchronous task with bulk input.
+    Users can call `task.results` and `task.errors` to obtain either the
+    job results upon success, or job errors upon failure.
+    """
+
+    def __init__(self, json):
+        super().__init__(json)
+
+    @property
+    @abstractmethod
+    def results(self):
+        ...
+
+    @property
+    @abstractmethod
+    def errors(self):
+        ...
+
+
+class DataRowImportTask(BulkTask):
+    """ Asynchronous task representing bulk data_row create job
+    """
+
+    task_type = TaskType.DATA_ROW_IMPORT
+
+    def __init__(self, json):
+        super().__init__(json)
+
+    @lru_cache()
+    def fetch_result(self) -> Dict[str, Any]:
+
+        def download_result():
+            response = requests.get(self.result_url)
+            response.raise_for_status()
+            return response.json()
+
+        if self.status == TaskStatus.IN_PROGRESS.name:
+            raise ResourceNotFoundError(
+                f"Task result for task '{self.id}' not found. Task is still in progress"
+            )
+        else:
+            return download_result()
+
+    # TODO: Need to paginate results
+    @property
+    def results(self) -> Union[Dict[str, Any], None]:
+        task_result = self.fetch_result()
+        if self.status == TaskStatus.FAILED.name or 'createdDataRows' not in task_result:
+            logger.warning(
+                "Task has failed. Please look at `task.errors` for more details"
+            )
+            return None
+
+        result = {}
+        result['created_data_rows'] = [
+            format_json_to_snake_case(dr)
+            for dr in task_result['createdDataRows']
+        ]
+        return result
+
+    @property
+    def errors(self) -> Union[Dict[str, Any], None]:
+        task_result = self.fetch_result()
+        # TODO: Improve error message from backend
+        errors = {}
+        if 'errors' in task_result and len(task_result['errors']) != 0:
+            errors['errors'] = task_result['errors']
+        if 'failedDataRows' in task_result:
+            errors['failed_data_rows'] = task_result['failedDataRows']
+
+        if errors:
+            return errors
+
+        return None
+
+    def wait_until_done(self, timeout_seconds=300) -> "TaskStatus":
+        assert self.id is not None
+        check_frequency = 2  # frequency of checking, in seconds
```

Comment:
also current SDK allows you to set check_frequency. Don't we want to preserve it and have like 10 seconds by default?

Reply:
I personally think customers don't need to worry about the check frequency.

Reply:
ah, true. it is only on label import. 
As soon as their requests don't add extra load on our servers - I'm fine with not having check_frequency configurable.